torchao/prototype/common/bitpacking.py

```diff
@@ -1,101 +1,133 @@
+    pad: if set to true, pads the dimension to be divisible by the scale
+    order: if set to true, packs elements such that the lower index elements occupy the most significant bits
     
     Returns: torch.Tensor - a tensor of packed elements.
     """
-    if by_rows:
-        return _pack_by_rows(data, container_size, data_size, device)
-    else:
-        return _pack_by_cols(data, container_size, data_size, device)   
     
-def _unpack_by_rows(data, data_size, device) -> torch.Tensor:
-    shape = data.shape
-    scale = data.element_size() * 8 // data_size
+    if element_type == "trinary":
+        data =  data + 1
+        
+    if container_dtype is not None:
+        data = data.to(container_dtype)
     
-    unpacked_data = torch.zeros((shape[0]*scale, *shape[1:]), dtype=data.dtype).to(device)
-    nbits = (1 << data_size) - 1 # mask for the last dtype_size bits
-    for i in range(scale):
-        shift_amt = data.element_size() * 8 - data_size * (i + 1) # how much to shift to get the ith uint
-        unpacked_data[i::scale] = ((data >> shift_amt) & (nbits))
-    return unpacked_data
+    container_size = torch.iinfo(data.dtype).bits
+    scale = container_size // element_bit_width
+    
+    if pad and data.shape[dim] % scale != 0:
+        padding = torch.zeros(mod_shape(data.shape, scale - data.shape[dim] % scale, dim), dtype=data.dtype).to(device)
+        data = torch.cat([data, padding], dim=dim).to(device)
+        
+    
+    torch._assert(data.shape[dim] >= scale, f"not enough values to pack along dimension {dim}")
+    torch._assert(data.shape[dim] % scale == 0, "size of pack dimension not divisble by scale")
+    return _pack(data, container_size, element_bit_width, scale, dim, order, device)
 
-def _unpack_by_cols(data, data_size) -> torch.Tensor:
-    shape = data.shape
-    scale = data.element_size() * 8 // data_size
-    unpacked_data = []
-    nbits = (1 << data_size) - 1 # mask for the last dtype_size bits
+
+
+def _pack(data, container_size, element_bit_width, scale, dim, order, device) -> torch.Tensor:
+    packed = torch.zeros(mod_shape(data.shape, data.shape[dim] // scale, dim), dtype=data.dtype).to(device)
+    slices = [slice(None)] * packed.ndim
     for i in range(scale):
-        shift_amt = data.element_size() * 8 - data_size * (i + 1) # how much to shift to get the ith uint
-        unpacked_data.append(((data >> shift_amt) & (nbits)).to(data.dtype))
-    return torch.stack(unpacked_data,dim=-1).view(*shape[:-1],shape[-1]*scale) # stack the unpacked data and reshape to the original shape
+        slices[dim] = slice(i, None, scale)
+        if order:
+            packed |= data[slices] << container_size-element_bit_width*(i+1)
+        else:
+            packed |= data[slices] << element_bit_width*i
+    return packed
 
-def _pack_by_rows(data, container_size, data_size, device) -> torch.Tensor:
+if __name__ == '__main__':
```

Comment:
Not a huge fan of having usage instructions in a python file. Ideally this should be a test instead or some usage example in a README in the `common/` folder